### ISSUE TITLE
CMS_add1_signer doesn't set signatureAlgorithm and it leads to fail with BIO_new_CMS scenario of cms creation.  

### ISSUE THREAD

**Issue by lo1ol** [openssl/openssl#18392]
openssl version 3.0.2

Hi! I noticed some problems of using CMS signing with BIO_new_CMS function. While I write data to returned bio I get an error from [here](https://github.com/openssl/openssl/blob/d649c51a5388912277dffb56d921eb720db54be1/crypto/asn1/tasn_enc.c#L374), while openssl tries to encode signature algorithm in Signer Info.

In my opinion the root of the problem is that signatureAlgorithm isn't set inside CMS_add1_signer function. What do you think about adding this functionality, for example like that:
```diff
diff --git a/crypto/cms/cms_sd.c b/crypto/cms/cms_sd.c
index 34c021bba6..3f730f4a2e 100644
--- a/crypto/cms/cms_sd.c
+++ b/crypto/cms/cms_sd.c
@@ -478,6 +478,10 @@ CMS_SignerInfo *CMS_add1_signer(CMS_ContentInfo *cms,
         }
     }

+    /* Set SignerInfo algorithm details if we used custom parameter */
+    if (si->pctx && !cms_sd_asn1_ctrl(si, 0))
+        goto err;
+
     if (!sd->signerInfos)
         sd->signerInfos = sk_CMS_SignerInfo_new_null();
     if (!sd->signerInfos || !sk_CMS_SignerInfo_push(sd->signerInfos, si))
```

Also I cut out problem part from my code to provide you small example for problem reproducing:
```c
    unsigned int flags = CMS_STREAM | CMS_BINARY;
    CMS_ContentInfo* cms = CMS_sign(signerCert, NULL, NULL, NULL, flags)
    if (!cms)
        return 1;

    // Here is signatureAlgorithm actually isn't set
    CMS_SignerInfo* si = CMS_add1_signer(cms, signerCert, pkey, NULL, flags | CMS_KEY_PARAM);
    if (!si)
        return 1;

    for (int i = 0; i < sk_X509_num(certs); ++i) {
        if (!CMS_add1_cert(cms, sk_X509_value(certs, i))) {
            auto err = ERR_get_error();
            if (ERR_LIB_CMS == ERR_GET_LIB(err) && CMS_R_CERTIFICATE_ALREADY_PRESENT == ERR_GET_REASON(err))
                continue;

            return 1;
        }
    }

    BIO* outputBio = BIO_new(BIO_s_mem());
    if (!outputBio)
        return 1;

     BIO* cmsBio = BIO_new_CMS(outputBio, cms);
    if (!cmsBio)
        return 1;

    while (dataLen) {
        // got an error. Of encoding to ASN.1 here
        int r = BIO_write(cmsBio, data, dataLen);
        if (r <= 0)
            return 1;
        data += r;
        dataLen -= r;
    }

    if (!BIO_flush(cmsBio))
        return 1;

    uint8_t* pCms = NULL;
    int cmsLen = 0;
    unsigned char * encodedCms = NULL;

    cmsLen = BIO_get_mem_data(outputBio, &pCms);
    if (cmsLen < 0)
            return 1;
    }
```

**Comment by t8m** [comment]
Is this a regression in 3.0? I.E., did it work for you on 1.1.1?

**Comment by lo1ol** [comment]
It worked at openssl 1.1.1 because [there is no check of returned value at this place](https://github.com/openssl/openssl/blob/d82e959e621a3d597f1e0d50ff8c2d8b96915fd7/crypto/asn1/tasn_enc.c#L360). But if it's were  there would be the error too.

**Comment by nhorman** [comment]
marking as inactive, to be closed at the end of 3.4 dev barring further input
